source/extensions/filters/listener/tls_inspector/tls_inspector.cc

```diff
@@ -228,15 +228,25 @@ std::vector<absl::string_view> Filter::getAlpnProtocols(const unsigned char* dat
     return protocols;
   }
 
-  absl::string_view str(reinterpret_cast<const char*>(data));
-  for (unsigned int i = 0; i < len;) {
-    uint32_t protocol_length = 0;
-    protocol_length <<= 8;
-    protocol_length |= data[i];
-    ++i;
-    absl::string_view protocol(str.substr(i, protocol_length));
-    protocols.push_back(protocol);
-    i += protocol_length;
+  // The data buffer contains a sequence of non-terminated strings, where each
+  // string is preceded by 1 byte that gives it's length (excluding the length
+  // byte itself). No string can be zero length, and the last string must not
+  // overrun or underun the buffer.
+  uint8_t protocol_length;
```

Comment:
It didn't look necessary to me because it does get initialised before use.

Reply:
Sorry if I'm a little dense, but where does protocol_length get initialised? 
(From what I can tell `i += protocol_length` is the first statement that mentions it?)

Reply:
The `i += protocol_length` statement doesn't get executed until after the first run through the loop, so the `protocol_length = data[i++]` statement is the first to be executed.

Reply:
evil :D

Reply:
Aaah yes, I missed that; thanks for explaining. Resolving this..

Reply:
though thinking about it, perhaps a one line comment about this won't hurt here just to make it obvious to future readers?

Reply:
or just initialize it... compiler might optimize it anyway, so...

We are also doing other evil stuff like changing the for loop variable in the middle of the loop. It might work, but my personal opinion is that in such cases we could just use while instead of for.

Reply:
So I mulled that over a bit, and now I agree on that I'd prefer a while loop as well. It seems like we're going through hoops to try and fit what we are doing into the original for loop construction, which I think started us off on the wrong foot: a while loop seems to be a more natural intuitive candidate here. @tedjpoole sorry about the flipflopping :-) curious to hear what you think?

Reply:
thanks for following up, I like the while loop better. Resolving
/cc @jwendell